### DISCUSSION TITLE
Guidance for SIngleton (per Server process)

### DISCUSSION THREAD

**Discussion by leowmjw** [remix-run/remix#1968] [General]
Trying to understand what is the suggested pattern for singleton (per Server process); to be used for connections that are considered heavyweight.

Let's say I have a scenario in the root level where it uses the loader to create a connection for Temporal Client to be used through the rest of this app; but it does not seem to behave as expected.

At root:
```
export const unstable_shouldReload = () => false;

// Because of above; load once below ..
export const loader = async () => {
    // This does not work ..
    const client = TemporalClientGlobal.getInstance()
..
```

When used in a route; let's say About page:
```
    try {
        // Expect to reuse connection ..
        const client = TemporalClientGlobal.getInstance()
        result =  await client.execute("example", {
            taskQueue: 'fetch-esm',
            workflowId: 'my-business-id',
            args: ['REMIX-LOAD'],
        });
        // DEBUG
        console.log(result); // Hello, Temporal!
...
```

The Output shows the connection is instantiated again.
```
New Heavyweight client needed!!
Hello, REMIX-LOAD!
GET /about?_data=routes%2Fabout 200 - - 1723.837 ms
New Heavyweight client needed!!
Hello, REMIX-LOAD!
GET /about?_data=routes%2Fabout 200 - - 783.111 ms

```

The singleton setup just uses the basic pattern;
```
class TemporalClientGlobal {
    private static _instance: WorkflowClient;

    private constructor() { }

    static getInstance() {
        if (this._instance) {
            return this._instance;
        }

        console.log("New Heavyweight client needed!!")
        const connection = new Connection({
                // address: "localhost:7233"
        })
        // Fails if gRPC conn not made in 1s
        connection.untilReady(1000)
        this._instance = new WorkflowClient(connection.service);

        return this._instance;
    }
}

export default TemporalClientGlobal;

```

Any guidance/advice or how to ensure the initialized instance is the one to be used instead of recreating?

**Comment by sergiodxa** [comment]
You can create your class in a file and export the instance, that should give you the singleton.

Note that, in development, Remix is clearing the require cache on each request and requiring the built code again so you don't need to restart the server on every file change, this will make your singleton be instantiated again, in development, one way to prevent this is to store the instance in `global` and read from there, if you check the Jokes tutorial it does that for Prisma.

In production it should work correctly as far as you keep the process running, this means if you use Serverless platforms like Vercel or Netlify or Cloudflare it will be impossible to avoid this. You need to use a deployment platform where you have the process running all the time.

**Reply by leowmjw** [reply]
@sergiodxa Thanks; that works! 

Thanks also for the good pointers; as I intend to deploy this to Cloudflare Workers; so it means there is no point to do this pattern as it will not be used.

**Comment by szhshp** [comment]
Hi @sergiodxa My site is on Vercel and the singleton are initialized on each visit.

I want to push my posts data to algolia on server started. I wrote it in `entry.server.tsx`. It doesn't make sense to run the push on each visit. Per you comments I believe this is a limitation on serverless pattern.

But is there any way that I can run specific job on `remix build` script?

**Reply by CanRau** [reply]
Hey have you figured out a way? If not, 1 option might be a resource route like /api/algolia, maybe even password protected, which you manually call or curl/fetch automated. Or you can have a separate script like `algolia.js` and in your build pipeline you run it `node algolia.js`.
